MobilityData/mobility-feed-api
# feat: search endpoint implementation

**Summary:**

Changes
- Add a materialized view for indexing feeds. Added all returned fields to the view to reduce DB calls.
- Implemented search endpoint querying the view with pattern '{query}:*', allowing incomplete lexemes in the queries.
- Refresh the materialized view after inserting data.

**Performance considerations**
We included all fields in the view to avoid multiple DB calls per HTTP request, which impacts the view's size. However, the current size is still small: ~1.6MB. We should monitor this size as the DB grows in the number of feeds.
DEV test was performed with the following results:
|    |  Time  |  Response Size  |  
|---|---|---|
|  limit=100, search_query=us | <150ms   |  90kb
|  limit=200, search_query=us  | <250ms   |  180kb
|  no limit, search_query=us  | ~1s   |  896kb
|  no limit, no search_query | ~1.6s   |  1.6MB


**Alternative solution considered**
We considered having a minimal footprint in the view to support the indexing and subsequence search action. Then, the Python code would call the DB multiple times to populate all remaining fields. Reducing the fields returned by the search endpoint is a possible optimization. 

Closes #371

**Expected behavior:** 

Explain and/or show screenshots for how you expect the pull request to work in your testing (in case other devices exhibit different behavior).

**Testing tips:**
Test locally the new search endpoint
- Update your local DB schema:
`docker-compose --env-file ./config/.env.local  up -d --force-recreate`
- Populate your DB by downloading the [csv catalog](https://bit.ly/catalogs-csv) and running the populate script
`./scripts/populate-db.sh {full-path of the catalog}`
- Start the Python application by using `./scripts/api-start.sh`
- Execute and play with different queries and parameters, example:
`curl --location 'http://localhost:8080/v1/search?search_query=mta&offset=0&limit=5' \
--header 'Accept: application/json'
`


Please make sure these boxes are checked before submitting your pull request - thanks!
